### ISSUE TITLE
directories are still traversed even if specified in --ignore 

### ISSUE THREAD

**Issue by mboynes** [squizlabs/PHP_CodeSniffer#933]
I have a project where phpcs takes a very long time to run. I traced the error down to an exhaustingly-large/deep `node_modules` directory, even though "node_modules" is in my `--ignore` paths. Turns out that every file within a directory is checked against the regular expressions.

If a directory is encountered in the traversal that matches an ignore pattern, that directory should not be traversed. Looks like this might be hard to do with the current `DirectoryIterator`/`RecursiveDirectoryIterator` implementation.

**Comment by mboynes** [comment]
I'm mistaken about the level of effort here; looks like `RecursiveCallbackFilterIterator` can make very light work of this. Here's a POC:

```php
<?php
$di = new RecursiveDirectoryIterator( './' );

$filtered_files = new RecursiveCallbackFilterIterator( $di, function( $current, $key, $iterator ) {
	// phpcs should probably do all of its filtering here
	if ( preg_match( '#node_modules#', $key ) ) {
		return false;
	}

	return true;
} );

$di = new RecursiveIteratorIterator(
	$filtered_files,
	0,
	RecursiveIteratorIterator::CATCH_GET_CHILD
);

foreach ( $di as $file ) {
	echo $file->getFileinfo() . "\n";
}
```

**Comment by gsherwood** [comment]
> I'm mistaken about the level of effort here

Actually, it's harder than you think because PHPCS 2.x supports PHP 5.1. I can't use RecursiveCallbackFilterIterator.

But I've actually done this already, but only in the 3.x version (the branch is called 3.0 and there are no releases yet). 3.x is a major rewrite of  pretty much everything, and bumps the min PHP version to 5.4. But I ended up using RecursiveFilterIterator there (I used RecursiveCallbackFilterIterator initially) because I was implementing a feature that allows you to have custom file filters.

I'm going to close this issue off and point you PR #620, which is where the discussion about the performance of iterating directories first started. It's also got my progress notes on it for the filtering change I described above. The commits referenced there (https://github.com/squizlabs/PHP_CodeSniffer/commit/4315292970fb1e48cf14b4a9012c6cf78f22df68 and https://github.com/squizlabs/PHP_CodeSniffer/commit/f90e08c0c284df4a6b368af7a1af0d19ce27853a) will show you how it was done.

But to summarise my position on this general idea: I won't be changing any directory iteration code in PHPCS 2.x because the risk is too great. But 3.x is the perfect place to make this sort of change, which has already been done. The issue you reported is a very real one, but I'm closing in on a first 3.x beta release so I need to be focusing on keeping 2.x stable.
